vhost-device-spi/src/main.rs

```diff
@@ -0,0 +1,173 @@
+// VIRTIO SPI Emulation via vhost-user
+//
+// Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+//          Haixu Cui <quic_haixcui@quicinc.com>
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+mod spi;
+mod vhu_spi;
+
+use log::error;
+use std::process::exit;
+use std::sync::{Arc, RwLock};
+use std::thread::{spawn, JoinHandle};
+
+use clap::Parser;
+use thiserror::Error as ThisError;
+use vhost_user_backend::VhostUserDaemon;
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+use spi::{PhysDevice, SpiController, SpiDevice};
+use vhu_spi::VhostUserSpiBackend;
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, ThisError)]
+/// Errors related to low level spi helpers
+pub(crate) enum Error {
+    #[error("Invalid socket count: {0}")]
+    SocketCountInvalid(usize),
+    #[error("Low level Spi failure: {0:?}")]
+    SpiFailure(spi::Error),
+    #[error("Could not create backend: {0}")]
+    CouldNotCreateBackend(vhu_spi::Error),
+    #[error("Could not create daemon: {0}")]
+    CouldNotCreateDaemon(vhost_user_backend::Error),
+    #[error("Fatal error: {0}")]
+    ServeFailed(vhost_user_backend::Error),
+}
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct SpiArgs {
+    /// Location of vhost-user Unix domain socket. This is suffixed by 0,1,2..socket_count-1.
+    #[clap(short, long)]
+    socket_path: String,
```

Comment:
What about `PathBuf`? (see https://github.com/rust-vmm/vhost-device/pull/446)

Reply:
Here I use PathBuf, and the socket path is "~/vspi.sock/0", not expected "~/vspi.sock0". I don't know if this case has a standard solution now.

Reply:
Sorry, I didn't get, can you elaborate a bit?

Reply:
@epilys Yes, the expected socket is `vspi.sock0` while I got `vspi.sock/0`. I don't understand how to edit the filename, could you please explain?

Reply:
@HaixuCui yes of course :)

Maybe this example will help:

https://github.com/rust-vmm/vhost-device/blob/e01d5e6bfe3850c62f725a7f606443426e3a8dc6/vhost-device-i2c/src/main.rs#L239-L255

It takes the filename as a separate variable, `socket_file_name`, adds the integer at the end of the string, then re-creates a path using the path and the new filename.

Otherwise, you can use <https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.set_file_name> directly if you have a `PathBuf`.

Reply:
Yes, it works and I can get the socket name correctly. I will update the `String `to `PathBuf `in next commit.

Reply:
Awesome! Path and PathBuf have weird APIs and it's easy to confuse them with strings, but once you get used to their concepts you'll have no problems anymore :)

Reply:
The basic concept is that a path is many "components" together. Directory Component, Separator, Directory Component, Separator, Filename.